File: Java/src/com/dbc/algorithm/Array/ContinuousSubarraySum.java
package com.dbc.algorithm.Array;

import java.util.HashMap;
import java.util.Map;

public class ContinuousSubarraySum {
    public static boolean checkSubarraySumEnum(int[] nums, int k) {
        if (nums.length < 2) {
            return false;
        }

        for (int i = 0; i < nums.length; i++) {
            int sum = nums[i];
            for (int j = i - 1; j >= 0; j--) {
                sum += nums[j];
                if (k == 0 && sum == 0) {
                    return true;
                } else if (k != 0 && sum % k == 0) {
                    return true;
                }
            }
        }

        return false;
    }

    public static boolean checkSubarraySumPrefix(int[] nums, int k) {
        int pre = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);

        for (int i = 0; i < nums.length; i++) {
            pre += nums[i];
            if (k != 0) {
                pre %= k;
            }
            if (i - map.getOrDefault(pre, i) > 1) {
                return true;
            }

            if (!map.containsKey(pre)) {
                map.put(pre, i);
            }
        }

        return false;
    }
}

File: core/tmp/Python/array/continuous_subarray_sum.py
def checkSubarraySum_violence(nums: list, k: int) -> bool:
    length = len(nums)
    if length < 2:
        return False

    for i in range(1, length):
        total = nums[i]
        for j in range(i - 1, -1, -1):
            total += nums[j]
            if k == 0 and total == 0:
                return True
            elif k != 0 and total % k == 0:
                return True

    return False


def checkSubarraySum_prefix(nums: list, k: int) -> bool:
    pre = 0
    mp = {0: -1}
    for i in range(len(nums)):
        pre += nums[i]
        if k != 0:
            pre %= k
        if i - mp.get(pre, i) > 1:
            return True
        
        if mp.get(pre, -2) == -2:
            mp[pre] = i

    return False


if __name__ == "__main__":
    print(checkSubarraySum_prefix(nums=[0,0], k = 0))
